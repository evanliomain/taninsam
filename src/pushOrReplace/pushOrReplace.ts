import { push } from '../push';

/**
 * @module array=>array
 */
/**
 * Push an element into an array, or replace an existing one if it matchs one.
 * @param replace Object with a match function and an element to replace in matching case.
 * @return the function to apply on the array to do perform the check, replace by the element if matching or push the element if not
 * @example
 * ```
 * pushOrReplace<number>({
 *       match: x => x%2 === 0,
 *       replacement: 2
 *    },{
 *       match: x => x%3 === 0,
 *       replacement: 3
 *    })([1, 2, 4]) // [1, 2, 2, 3]
 * ```
 * @example Using the chain
 * ```
 * chain([1, 2, 4])
 *   .chain(pushOrReplace<number>({
 *       match: x => x%2 === 0,
 *       replacement: 2
 *    },{
 *       match: x => x%3 === 0,
 *       replacement: 3
 *    }))
 *   .value() // [1, 2, 2, 3]
 * ```
 */
export function pushOrReplace<T>(
  ...replace: {
    match: (element: T) => boolean;
    replacement: T;
  }[]
): (array: ReadonlyArray<T>) => ReadonlyArray<T> {
  return (array: ReadonlyArray<T>) =>
    replace.reduce(
      (accuList, r) =>
        accuList.every(element => !r.match(element))
          ? push(r.replacement)(accuList)
          : accuList.map(
              (element: T) => (r.match(element) ? r.replacement : element)
            ),
      array
    );
}
