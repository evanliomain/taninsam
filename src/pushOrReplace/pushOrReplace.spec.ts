import { pushOrReplace } from './pushOrReplace';

describe('pushOrReplace function', () => {
  const input = [1, 2, 4];

  test('is a pure function', () => {
    pushOrReplace({
      match: x => x % 2 === 0,
      replacement: 2
    })(input);
    expect(input).toMatchSnapshot();
  });

  test(
    '[1, 2, 4] ' +
      '|> pushOrReplace({ match: x => x % 2 === 0, replacement: 2 }, { match: x => x % 3 === 0, replacement: 3 })' +
      ' === [1, 2, 2, 3]',
    () => {
      expect(
        pushOrReplace<number>(
          {
            match: x => x % 2 === 0,
            replacement: 2
          },
          {
            match: x => x % 3 === 0,
            replacement: 3
          }
        )(input)
      ).toMatchSnapshot();
    }
  );

  test('[] |> pushOrReplace({ match: x => x % 2 === 0, replacement: 2 }) === [2]', () => {
    expect(
      pushOrReplace<number>({
        match: x => x % 2 === 0,
        replacement: 2
      })([])
    ).toMatchSnapshot();
  });
});
