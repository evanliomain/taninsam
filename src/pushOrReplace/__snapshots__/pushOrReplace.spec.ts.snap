// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pushOrReplace function [] |> pushOrReplace({ match: x => x % 2 === 0, replacement: 2 }) === [2] 1`] = `
Array [
  2,
]
`;

exports[`pushOrReplace function [1, 2, 4] |> pushOrReplace({ match: x => x % 2 === 0, replacement: 2 }, { match: x => x % 3 === 0, replacement: 3 }) === [1, 2, 2, 3] 1`] = `
Array [
  1,
  2,
  2,
  3,
]
`;

exports[`pushOrReplace function [4, 6, 8] |> pushOrReplace({ match: x => x % 2 === 0, replacement: 2 }) === [2, 2, 2] 1`] = `
Array [
  2,
  2,
  2,
]
`;

exports[`pushOrReplace function is a pure function 1`] = `
Array [
  1,
  2,
  4,
]
`;
